polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -0,0 +1,1047 @@
+	pub candidate_hash: CandidateHash,
+	cumulative_modifications: ConstraintModifications,
+}
+
+impl FragmentNode {
+	fn relay_parent(&self) -> Hash {
+		self.fragment.relay_parent().hash
+	}
+}
+
+/// This is a chain of candidates based on some underlying storage of candidates and a scope.
+///
+/// All nodes in the chain must be either pending availability or within the scope. Within the scope
+/// means it's built off of the relay-parent or an ancestor.
+pub(crate) struct FragmentChain {
+	scope: Scope,
+
+	pub chain: Vec<FragmentNode>,
+
+	candidates: HashSet<CandidateHash>,
+
+	// Index from head data hash to candidate hashes with that head data as a parent.
+	by_parent_head: HashMap<Hash, CandidateHash>,
+	// Index from head data hash to candidate hashes outputting that head data.
+	by_output_head: HashMap<Hash, CandidateHash>,
+}
+
+impl FragmentChain {
+	/// Create a new [`FragmentChain`] with given scope and populated from the storage.
+	pub fn populate(scope: Scope, storage: &CandidateStorage) -> Self {
+		gum::trace!(
+			target: LOG_TARGET,
+			relay_parent = ?scope.relay_parent.hash,
+			relay_parent_num = scope.relay_parent.number,
+			para_id = ?scope.para,
+			ancestors = scope.ancestors.len(),
+			"Instantiating Fragment Chain",
+		);
+
+		let mut fragment_chain = Self {
+			scope,
+			chain: Vec::new(),
+			candidates: HashSet::new(),
+			by_parent_head: HashMap::new(),
+			by_output_head: HashMap::new(),
+		};
+
+		fragment_chain.populate_chain(storage);
+
+		fragment_chain
+	}
+
+	/// Get the scope of the Fragment Chain.
+	pub fn scope(&self) -> &Scope {
+		&self.scope
+	}
+
+	/// Returns an O(n) iterator over the hashes of candidates contained in the
+	/// chain.
+	pub(crate) fn candidates(&self) -> impl Iterator<Item = CandidateHash> + '_ {
+		self.candidates.iter().cloned()
+	}
+
+	/// Returns the number of candidates in the chain
+	pub(crate) fn len(&self) -> usize {
+		self.candidates.len()
+	}
+
+	/// Whether the candidate exists.
+	pub(crate) fn contains_candidate(&self, candidate: &CandidateHash) -> bool {
+		self.candidates.contains(candidate)
+	}
+
+	/// Try accumulating more candidates onto the chain.
+	///
+	/// Candidates can only be added if they build on the already existing chain.
+	pub(crate) fn repopulate(&mut self, storage: &CandidateStorage) {
+		self.populate_chain(storage);
+	}
+
+	/// Returns the hypothetical state of a candidate with the given hash and parent head data
+	/// in regards to the existing chain.
+	///
+	/// Returns true if either:
+	/// - the candidate is already present
+	/// - the candidate can be added to the chain
+	/// - the candidate could potentially be added to the chain in the future (its ancestors are
+	///   still unknown but it doesn't violate other rules).
+	///
+	/// If this returns false, the candidate could never be added to the current chain (not now, not
+	/// ever)
+	pub(crate) fn hypothetical_membership(
+		&self,
+		candidate_hash: CandidateHash,
```

Comment:
It feels weird to require the candidate hash passed separately, it should be known in `HypotheticalCandidate`